site/blog/updates/why-sql-for-infrastructure.mdx

```diff
@@ -102,8 +102,69 @@ Infrastructure management *generally* is not that straightforward. Changes to in
 So declarative programming for infrastructure, also known as [Infrastructure as Code](https://en.wikipedia.org/wiki/Infrastructure_as_code), must take the existing state and generate the imperative operations to perform based on the differences between the existing state and the declared desired state. This is like a contracting company being given an architecture draft for a new restaurant and being told to remodel an existing commercial space (maybe it was also a restaurant, maybe it was a clothing store, etc) and the contracting company figuring out what needs to be torn down first, what can be re-used, and what needs to be built new.
 
 ```hcl
-resource "ssh_accessible_ec2_instance" {
-  implementation = "tbd"
+terraform {
+  required_providers {
+    aws = {
+      source = "hashicorp/aws"
+      version = "4.60.0"
+    }
+  }
+}
+
+data "external" "example" {
+  program = [
+    "bash",
+    "-c",
+    <<-EOT
+      if [ ! -f ./login.pem ]; then
+        # Terraform doesn't support creating the PEM file via AWS. It has to be
+        # created locally and then imported into AWS. Inside of a conditional so
+        # we don't accidentally re-create the file between 'plan' and 'apply'
+        yes '' | ssh-keygen -t ed25519 -m PEM -f login.pem
+      fi
+      chmod 600 login.pem
+      # The output of this script needs to be JSON formatted. This is a bit wonky
+      # but building it over time via 'jq' operators is actually harder to follow
+      echo {
+      echo   '"access_key_id": "'$(aws configure get aws_access_key_id)'",'
+      echo   '"secret_access_key": "'$(aws configure get aws_secret_access_key)'",'
+      echo   '"public_key": "'$(cat login.pem.pub)'"'
```

Comment:
I think Terraform will use the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` environment variables. Also, for the key we can use [`local_file`](https://registry.terraform.io/providers/hashicorp/local/latest/docs/data-sources/file) instead?

Reply:
Yes, but I didn't like the not-self-contained nature of that approach versus this one for both situations. It requires us to invoke `terraform plan` like `AWS_ACCESS_KEY_ID=$(aws configure get aws_access_key_id) AWS_SECRET_ACCESS_KEY=$(aws configure get aws_secret_access_key) terraform plan` if we don't want to make the assumption that those environment variables were already set.

Also, I hate environment variables because they are common footguns when trying to reproduce what some piece of code does, and I prefer to minimize them as much as possible.

Reply:
I can add a comment into the Terraform code that the shell script isn't necessary if you set the appropriate environment variables outside and create the ssh key ahead of time and just reference it here, but those would be manual steps you have to ensure you do before running the rest of the Terraform code which makes actually executing the example more difficult.